### PR TITLE
Fixes #RHIROS-786 - handle exception on invalid value for rating

### PR DESCRIPTION
**Pull request by kgaikwad**
With this commit, server will return proper error code & message instead of 500 server error on submitting invalid value against rating via /recommendation_ratings POST API request.

With latest changes, additionally handled below scenarios:
- when user sends invalid JSON or incorrect JSON format 
- when user sends invalid inventory_id

### CODE REVIEW

File: ros/api/v1/recommendation_ratings.py [hunk 1]
```diff
@@ -28,20 +28,28 @@ def post(self):
...
+                f'Possible values - { *allowed_choices, }'
+            )
+        rating = None
+        try:
+            rating = int(data['rating'])
+        except ValueError:
+            abort(400, message=(invalid_value_msg(data['rating'])))
+
+        if rating not in allowed_choices:
+            abort(422, message=(invalid_value_msg(data['rating'])))
```
Comment (kgaikwad) [inline ros/api/v1/recommendation_ratings.py:45]: Please let me know your thoughts on this `422` error code. Do we want to keep it as it is or you want me to change.

Comment (upadhyeammit) [reply ros/api/v1/recommendation_ratings.py:45]: As per the docs : https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422 the 422 is good fit here!

**Review by upadhyeammit — APPROVED** [review]
LGTM!

**Review by r14chandra — APPROVED** [review]
looks good

File: ros/api/v1/recommendation_ratings.py [hunk 2]
```diff
@@ -28,20 +28,28 @@ def post(self):
...
         inventory_id = data['inventory_id']
-        rating = int(data['rating'])
-        if rating not in [c.value for c in RatingChoicesEnum]:
-            abort(
-                422,
-                message=(
-                    "{} is not a valid value for rating".format(rating)))
+        allowed_choices = [c.value for c in RatingChoicesEnum]
+
+        def invalid_value_msg(val):
```
Comment (saltgen) [inline ros/api/v1/recommendation_ratings.py:33]: Looks like this is a good contender for decorator validation. 

We can abstract all this code into a nice little function and keep our code pretty and modular.

Just like how we use ` @marshal_with(x_fields)`

Example:

```
def decorator(func):
    def inner():
        print("run something before actual function")
        func()
        print("run something after actual function")
    return inner


@decorator
def func():
    print("something...")


func()
"""
run something before actual function
something...
run something after actual function
"""
```

References:  

- https://levelup.gitconnected.com/validations-in-python-using-metaprogramming-and-decorators-advanced-python-ee4d4278a6b3

- https://stackoverflow.com/questions/50711796/how-to-validate-arguments-using-decorators-in-python

File: ros/api/v1/recommendation_ratings.py [hunk 3]
```diff
@@ -28,20 +28,28 @@ def post(self):
...
+            return (
+                f"'{val}' is invalid value for rating."
+                f'Possible values - { *allowed_choices, }'
```
Comment (saltgen) [inline ros/api/v1/recommendation_ratings.py:36]: Nitpick :

Let's have consistent surrounding quotes?

File: ros/openapi/openapi.json [hunk 4]
```diff
@@ -259,6 +259,26 @@
...
+                        "description": "System doesn't exist",
+                        "content": {
+                            "application/json": {
+                                "schema": {
+                                    "$ref": "#/components/schemas/Message"
+                                }
+                            }
+                        }
+                    },
+                    "400": {
```
Comment (saltgen) [inline ros/openapi/openapi.json:273]: Should this be 422?

Comment (kgaikwad) [reply ros/openapi/openapi.json:273]: Details about 422 already present above.

File: ros/api/v1/recommendation_ratings.py (hunk 3 above)
Comment (kgaikwad) [reply ros/api/v1/recommendation_ratings.py:36]: @saltgen - In above sentence,  given value for rating is enclosed in single quotes (') so used ("") double quotes.
Did you mean to add double quotes on this line-36, correct?

Comment (saltgen) [reply ros/api/v1/recommendation_ratings.py:36]: Yes, L36, correct

File: ros/openapi/openapi.json (hunk 4 above)
Comment (saltgen) [reply ros/openapi/openapi.json:273]: Ah okay, in that case shall we update the desc. L274?

Comment (saltgen) [reply ros/openapi/openapi.json:273]: L254 suggestion: _Invalid choice of input for rating_, as discussed L274 looks good

File: ros/api/v1/recommendation_ratings.py (hunk 2 above)
Comment (kgaikwad) [reply ros/api/v1/recommendation_ratings.py]: Good idea. Added changes, please request to review.

File: ros/api/v1/recommendation_ratings.py [hunk 5]
```diff
@@ -5,46 +5,90 @@
...
-        'rating': fields.Integer
-    }
+    def error_msg(error_code, value):
+        errors = {
+            400: "is invalid value for rating.",
+            422: "is invalid choice of input for rating."
+        }
+        return (
+            f"'{value}' { errors.get(error_code, 'Invalid') }"
+            f"Possible values - { *allowed_choices, }"
```
Comment (saltgen) [inline ros/api/v1/recommendation_ratings.py:19]:
```suggestion
            f" Possible values - { *allowed_choices, }"
```

File: ros/api/v1/recommendation_ratings.py [hunk 6]
```diff
@@ -5,46 +5,90 @@
...
+        """
+        username = kwargs.get('username')
+        inventory_id = kwargs.get('inventory_id')
+        rating = kwargs.get('rating')
+        system_id = kwargs.get('system_id')
 
         rating_record = RecommendationRating.query.filter(
-            RecommendationRating.system_id == system.id,
+            RecommendationRating.system_id == system_id,
             RecommendationRating.rated_by == username).first()
```
Comment (saltgen) [inline ros/api/v1/recommendation_ratings.py:92]:
```suggestion
            RecommendationRating.rated_by == username
        ).first()
```

File: ros/api/v1/recommendation_ratings.py (hunk 5 above)
Comment (saltgen) [reply ros/api/v1/recommendation_ratings.py:19]: Or, shall we remove L19?
The openapi.json has an enum with a list of choices

File: ros/api/v1/recommendation_ratings.py [hunk 7]
```diff
@@ -5,46 +5,90 @@
...
         if system is None:
-            abort(404, message="System {} doesn't exist"
-                  .format(inventory_id))
+            abort(404, message=f"System {inventory_id} doesn't exist")
+
+        return system.id
+
+    def validate_request(*args, **kwargs):
+        username = check_for_user()
+        data = json.loads(request.data)
```
Comment (saltgen) [inline ros/api/v1/recommendation_ratings.py:56]: Tested w/o `inventory_id` and `rating`, getting a 500 server error on this line.

**Error:** json.decoder.JSONDecodeError: Expecting value

Happy path tests went fine :+1:

File: ros/api/v1/recommendation_ratings.py [hunk 8]
```diff
@@ -5,46 +5,90 @@
 from ros.lib.utils import identity, user_data_from_identity
 
 
-class RecommendationRatingsApi(Resource):
+def validate_rating_post_api(func):
```
Comment (saltgen) [inline ros/api/v1/recommendation_ratings.py:8]: Shall we move this decorator to utils as of now?
If we come up with more later, we can create a dedicated file.

**Review by saltgen — CHANGES_REQUESTED** [review]
Appreciate the help with the validation decorator! :raised_hands:
I have tested the changes locally.
Left a few comments and additionally would like to suggest a happy and sad path unittest.

File: ros/api/v1/recommendation_ratings.py (hunk 5 above)
Comment (kgaikwad) [reply ros/api/v1/recommendation_ratings.py:19]: I would prefer to keep it in error response.

File: ros/api/v1/recommendation_ratings.py (hunk 7 above)
Comment (kgaikwad) [reply ros/api/v1/recommendation_ratings.py]: Handle these exceptions too. Please request to review and confirm the changes.

File: ros/api/common/utils.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+import logging
+from flask import request
+from flask_restful import abort
+from ros.lib.models import System, RatingChoicesEnum
+from ros.lib.utils import identity, user_data_from_identity
+
+LOG = logging.getLogger(__name__)
+
+
+def validate_rating_post_api(func):
```
Comment (saltgen) [inline ros/api/common/utils.py:11]:
```suggestion
def validate_rating_data(func):
```

File: ros/api/common/utils.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        ident = identity(request)['identity']
+        user = user_data_from_identity(ident)
+        username = user['username'] if 'username' in user else None
+
+        if username is None:
+            abort(403, message="Username doesn't exist")
+
+        return username
+
+    def check_for_system(inventory_id):
```
Comment (saltgen) [inline ros/api/common/utils.py:47]: An invalid `inventory_id` leads to a 500 error, I have checked the same for `rating` and it works :+1: 

```
api-post-hosts:
	curl  -v -H "Content-Type: application/json"  -H "x-rh-identity: ${b64_identity}" \
 		  -X POST -d '{ "inventory_id": "hello-world", "rating": 1}' \
          "localhost:8005/api/ros/v1/rating" | python -m json.tool
```

**Review by saltgen — CHANGES_REQUESTED** [review]

**Comment by kgaikwad** [comment]
> Appreciate the help with the validation decorator! raised_hands I have tested the changes locally. Left a few comments and additionally would like to suggest a happy and sad path unittest.

@saltgen  +1, added more unitests related to POST rating request to cover sad path scenarios. Please request to review and let me know if any suggestions.

**Review by saltgen — APPROVED** [review]
Good stuff!

**Pull request merged**